### PR TITLE
[SR-10979] Add fix-it for missing operator

### PR DESCRIPTION
**Pull request by mcichecki**
This PR adds fix-it for operator functions when operator is not declared.

Resolves [SR-10979](https://bugs.swift.org/browse/SR-10979).

### CODE REVIEW

File: lib/Sema/TypeCheckDecl.cpp [hunk 1]
```diff
@@ -3641,8 +3641,37 @@ void bindFuncDeclToOperator(TypeChecker &TC, FuncDecl *FD) {
...
+    if (dyn_cast<SourceFile>(FD->getParent())) {
+      // Parent context is SourceFile, insertion location is start of func declaration
+      // or unary operator
+      insertionLoc = FD->isUnaryOperator() ? FD->getAttrs().getStartLoc() : FD->getStartLoc();
+    } else {
+      // Finding top-level decl context before SourceFile and inserting before it
+      for (DeclContext *CurContext = FD->getLocalContext();
+           CurContext;
+           CurContext = CurContext->getParent()) {
+        insertionLoc = CurContext->getAsDecl()->getStartLoc();
```
Comment (harlanhaskins) [inline lib/Sema/TypeCheckDecl.cpp]: Seems like this should be inside the `if`.

File: lib/Sema/TypeCheckDecl.cpp [hunk 2]
```diff
@@ -3641,8 +3641,37 @@ void bindFuncDeclToOperator(TypeChecker &TC, FuncDecl *FD) {
...
+      else
+        insertion += "postfix operator ";
+    } else if (numOfParams == 2) {
+      insertion += "infix operator ";
+    }
+
+    insertion += operatorName.str();
+    insertion += " : <# Precedence Group #>\n";
+    TC.diagnose(FD, diag::declared_operator_without_operator_decl)
+    .fixItInsert(insertionLoc, insertion);
```
Comment (harlanhaskins) [inline lib/Sema/TypeCheckDecl.cpp]: Can you guard this on if `insertionLoc.isValid()`?

**Review by harlanhaskins — APPROVED** [review]
This looks good to me, thanks for fixing this! I have a couple of suggestions, but overall this looks goof.

File: lib/Sema/TypeCheckDecl.cpp [hunk 3]
```diff
@@ -3641,8 +3641,37 @@ void bindFuncDeclToOperator(TypeChecker &TC, FuncDecl *FD) {
   }
 
   if (!op) {
-    // FIXME: Add Fix-It introducing an operator declaration?
-    TC.diagnose(FD, diag::declared_operator_without_operator_decl);
+    SourceLoc insertionLoc;
+    if (dyn_cast<SourceFile>(FD->getParent())) {
```
Comment (theblixguy) [inline lib/Sema/TypeCheckDecl.cpp]:
```suggestion
    if (isa<SourceFile>(FD->getParent())) {
```

File: lib/Sema/TypeCheckDecl.cpp [hunk 4]
```diff
@@ -3641,8 +3641,37 @@ void bindFuncDeclToOperator(TypeChecker &TC, FuncDecl *FD) {
...
+      // Parent context is SourceFile, insertion location is start of func declaration
+      // or unary operator
+      insertionLoc = FD->isUnaryOperator() ? FD->getAttrs().getStartLoc() : FD->getStartLoc();
+    } else {
+      // Finding top-level decl context before SourceFile and inserting before it
+      for (DeclContext *CurContext = FD->getLocalContext();
+           CurContext;
+           CurContext = CurContext->getParent()) {
+        insertionLoc = CurContext->getAsDecl()->getStartLoc();
+        if (dyn_cast<SourceFile>(CurContext->getParent()))
```
Comment (theblixguy) [inline lib/Sema/TypeCheckDecl.cpp]:
```suggestion
        if (isa<SourceFile>(CurContext->getParent()))
```

**Review by theblixguy — COMMENTED** [review]
Just one small suggestion

File: lib/Sema/TypeCheckDecl.cpp (hunk 1 above)
Comment (mcichecki) [reply lib/Sema/TypeCheckDecl.cpp]: I actually moved the condition to the for loop. I think that's the cleanest way to get the last context that is not sourcefile. Hope that's ok with you 🙂

File: lib/Sema/TypeCheckDecl.cpp (hunk 3 above)
Comment (mcichecki) [reply lib/Sema/TypeCheckDecl.cpp]: Thanks for the suggestion, applied.

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Comment by harlanhaskins** [comment]
⛵

Thanks @mcichecki! And welcome to the Swift project, we really appreciate your contribution!

**Pull request merged**